draft-robert-mls-extensions.md

```diff
@@ -93,6 +102,173 @@ to recognize which messages contain AppAcks.  The application can also have
+only sent to specific group members, either for privacy or for efficiency
+reasons.
+
+Targeted messages are a way to achieve this without having to create a new group
+with the sender and the specific recipients – which might not be possible or
+desired. Instead, targeted messages define the format and encryption of a
+message that is sent from a member of an existing group to another member of
+that group.
+
+The goal is to provide a one-shot messaging mechanism that provides
+confidentiality and authentication.
+
+### Format
+
+This extensions extens the MLS protocol to include a new message type,
+`TargetedMessage` in `WireFormat` and `MLSMessage`:
+
+~~~ tls
+enum {
+  ...
+  mls_targeted_message(6),
+  ...
+  (255)
+} WireFormat;
+
+struct {
+    ProtocolVersion version = mls10;
+    WireFormat wire_format;
+    select (MLSMessage.wire_format) {
+        ...
+        case mls_targeted_message:
+            TargetedMessage targeted_message;
+    }
+} MLSMessage;
+~~~
+
+The `TargetedMessage` message type is defined as follows:
+
+~~~ tls
+struct {
+  opaque group_id<V>;
+  uint64 epoch;
+  uint32 recipient_leaf_index;
+  opaque authenticated_data<V>;
+  opaque encrypted_sender_data<V>;
+  opaque encrypted_targeted_message_content<V>;
+} TargetedMessage;
+
+enum {
+  HPKEAuth,
+  Signature,
+} TargetedMessageAuthScheme;
+
+struct {
+  TargetedMessageAuthScheme authentication_scheme;
+  select (authentication_scheme) {
+    case HPKEAuth:
+    case Signature:
+      opaque signature<V>;
+  }
+  HPKECiphertext ciphertext;
+} TargetedMessageContent;
+
+struct {
+  opaque group_id<V>;
+  uint64 epoch;
+  uint32 recipient_leaf_index;
+  opaque authenticated_data<V>;
+  opaque encrypted_sender_data<V>;
+  TargetedMessageAuthScheme authentication_scheme;
+} TargetedMessageTBM;
+
+struct {
+  TargetedMessageTBM targeted_message_tbm;
+  HPKECiphertext hpke_ciphertext;
+} TargetedMessageTBS;
+
+struct {
+  opaque group_id<V>;
+  uint64 epoch;
+  opaque label<V> = "MLS 1.0 targeted message psk";
+} PSKId;
+~~~
+
+### Encryption
+
+Targeted messages use HPKE to encrypt the message content between two leaves.
+The HPKE keys of the `LeafNode` are used to that effect, namely the
+`encryption_key` field.
+
+In addition, the sender data encryption from section 7.3.2 {{mls-protocol}} is
+used to encrypt `MLSSenderData`. `MLSSenderData.leaf_index` is the leaf index of
+the sender. The `MLSSenderData.generation` field is not used and MUST be set to
+0.
+
+### Authentication
+
+For ciphersuites that support it, HPKE `mode_auth_psk` is used for
+authentication. For other ciphersuites, HPKE `mode_psk` is used along with a
+signature. The authentication scheme is indicated by the `authentication_scheme`
+field in `TargetedMessageContent`. See {{guidance-on-authentication-schemes}}
+for more information.
+
+For the PSK part of the authentication, clients export a dedicated secret:
+
+```
+targeted_message_psk = MLS-Exporter("targeted message psk", "", KDF.Nh)
```

Comment:
Is the idea that the PSK binds the message to an epoch? The security properties are a little unclear because the HPKE key is a relatively long-lived key but this one is not.
Reading further it looks like you really want to use PSK (encryption) and want use HPKE/Signature (authentication) only for authentication? This doesn't become clear from the text. I thought you were doing HPKE (encryption) + hpke_auth/signature (authentication). But then the encryption is always HPKE (not PSK), which makes me very confused.

Reply:
I'm all for using PSK here 🙂 
My main point is that it needs to become clear where the security is coming from and how the different things work together.
The PSK in HPKE is only used for the key schedule, not the KEM. Hence, the encapsulation is independent of the PSK and only dependent on the long-lived HPKE key. The encryption happens with the key from the HPKE key schedule that includes the PSK.

So I'd suggest to add your explanations about the security guarantees and some basic explanation (like the one above) for how the PSK is used in HPKE (to make it clear where the security properties come from).